File: C++/balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool result;
    
    bool isBalanced(TreeNode* root) {
        result = true;
        checkDepth(root, 0);
        return result;
    }
    
    int checkDepth(TreeNode* root, int depth) {
        if (root) {
            int a = checkDepth(root->left, depth + 1);
            int b = checkDepth(root->right, depth + 1);
            if (abs(a - b) < 2) {
                return max(a, b);
            } else {
                result = false;
                return depth;
            }
        }
        else {
            return depth;
        }
    }
};

File: Kotlin/balanced-binary-tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun isBalanced(root: TreeNode?): Boolean {
        if (root == null) {
            return true
        }
        if (!isBalanced(root?.left)) {
            return false
        }
        if (!isBalanced(root?.right)) {
            return false
        }        
        var leftHeight = treeHeight(root?.left)
        var rightHeight = treeHeight(root?.right)
        var diff = kotlin.math.abs(leftHeight - rightHeight)
        return diff < 2
    }

    fun treeHeight(root: TreeNode?): Int {
        if (root != null) {
            var left = treeHeight(root?.left)
            var right = treeHeight(root?.right)
            return 1 + kotlin.math.max(left, right)
        }
        return 0
    }
}